community/primitive-collections/src/main/java/org/neo4j/concurrent/AsyncEvents.java

```diff
@@ -0,0 +1,162 @@
+ * foreground as a fall-back.
+ *
+ * Note, however, that no events are processed until the background thread is started.
+ *
+ * The {@code AsyncEvents} is given a {@link Consumer} of the specified event type upon construction, and will use it
+ * for doing the actual processing of events once they have been collected.
+ *
+ * @param <T> The type of events the {@code AsyncEvents} will process.
+ */
+public class AsyncEvents<T extends AsyncEvent> implements AsyncEventSender<T>, Runnable
+{
+    // TODO use VarHandles in Java 9
+    private static AtomicReferenceFieldUpdater<AsyncEvents,AsyncEvent> updater =
+            AtomicReferenceFieldUpdater.newUpdater( AsyncEvents.class, AsyncEvent.class, "stack" );
+
+    private final EndEvent endSentinel;
+    private final ShutdownEvent shutdownSentinel;
+    private final Consumer<T> eventConsumer;
+
+    @SuppressWarnings( {"unused", "FieldCanBeLocal"} )
+    private volatile AsyncEvent stack; // Accessed via AtomicReferenceFieldUpdater
+    private volatile boolean shutdown;
+
+    /**
+     * Construct a new {@code AsyncEvents} instance, that will use the given consumer to process the events.
+     *
+     * @param eventConsumer The {@link Consumer} used for processing the events that are sent in.
+     */
+    public AsyncEvents( Consumer<T> eventConsumer )
+    {
+        endSentinel = new EndEvent();
+        shutdownSentinel = new ShutdownEvent();
+        this.eventConsumer = eventConsumer;
+        stack = endSentinel;
+    }
+
+    @Override
+    public void send( T event )
+    {
+        AsyncEvent prev = updater.getAndSet( this, event );
+        assert prev != null;
+        event.next = prev;
+        if ( prev == endSentinel )
+        {
+            LockSupport.unpark( endSentinel.runner );
+        }
+        else if ( prev == shutdownSentinel )
+        {
+            AsyncEvent events = updater.getAndSet( this, shutdownSentinel );
+            process( events );
+        }
+    }
+
+    @Override
+    public void run()
+    {
+        endSentinel.runner = Thread.currentThread();
+
+        do
+        {
+            AsyncEvent events = updater.getAndSet( this, endSentinel );
+            process( events );
+
+            if ( stack == endSentinel && !shutdown )
+            {
+                LockSupport.park( this );
+            }
+        }
+        while ( !shutdown );
+
+        AsyncEvent events = updater.getAndSet( this, shutdownSentinel );
+        process( events );
+    }
+
+    private void process( AsyncEvent events )
+    {
+        events = reverseAndStripEndMark( events );
+
+        while ( events != null )
+        {
+            @SuppressWarnings( "unchecked" )
+            T event = (T) events;
+            eventConsumer.accept( event );
+            events = events.next;
+        }
+
+        endSentinel.next = null;
+        shutdownSentinel.next = null;
+    }
+
+    private AsyncEvent reverseAndStripEndMark( AsyncEvent events )
```

Comment:
Because a stack implies fewer pointers to maintain invariants on, making the code for adding events simpler and faster.